### DISCUSSION TITLE
Enable pointers to be readonly

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/csharplang#5743] [Language Ideas] [answered]
I know there have been a few discussion #955. In my situation I need the possibility to have over `void*` pointer as readonly.
The C++ syntax as described by @tannergooding in #955 is more or less exactly what I need. To show an example:

```cs
 /// <summary>
    /// Structure that describes a memory pointer.
    /// </summary>
    public interface IReadOnlyMemoryHandle : IDisposable
    {
        /// <summary>
        /// Gets the data type the pointer contains.
        /// </summary>
        public Type DataType { get; }

        /// <summary>
        /// Gets the size (in bytes) of one element.
        /// </summary>
        public int SizeOf { get; }

        /// <summary>
        /// Gets the number of components for each element.
        /// </summary>
        /// <remarks>
        /// This is relevant for data that consists of multiple components 
        /// (like complex:2 or color RGB:3 values)
        /// </remarks>
        public int NoOfComponents { get; }

        /// <summary>
        /// Gets the pointer.
        /// </summary>
        /// <returns></returns>
        public unsafe void* GetPointer();
    }
```

I call it IReadOnlyMemoryHandle but I cannot make the pointer readonly. The consumer of this struct should only be able to read the pointer but should not be allowed to write it.

The question is if there is a C# way of how to do it. Analyzers are not what I meant...

**Comment by HaloFour** [comment, marked as answer]
Why not expose it as a `ReadOnlySpan<byte>` instead?

**Reply by msedi** [reply]
@HaloFour : Problem is that it's part of a structure with an interface and I cannot make it a ref struct. Additionally I cannot make it a Memory<T> because the data is partly unmanaged memory from memory mapped files or from CUDA managed memory, so void * is there only common denominator that works outside the tight Span<T> context. 

But if Span<T> could be used, then I would agree.

**Reply by Joe4evr** [reply]
You don't need a ref struct to *return* a Span. You can store the pointer and create a span around it:
```cs
public unsafe struct Impl : IReadOnlyMemoryHandle
{
    //....
    private void* _ptr;

    public ReadOnlySpan<byte> GetPointer() => new ReadOnlySpan<byte>(_ptr, SizeOf * NoOfElements);
}
```

**Reply by tannergooding** [reply]
Exposing a `ReadOnlySpan` is not always possible or desirable depending on context, perf requirements, or usage scenario.

That being said, pointers are already unsafe and I'm not convinced that updating the compiler to provide additional safety for an already niche scenario is worth it in the end when we do have all the other ways of tracking or providing additional safety on-top.

It would definitely be a nice feature to have (for those of us that use pointers frequently) and to be able to correctly expose the mutability information that native has; but there are also so many other features that I'd rather have above this that seeing it bubble up high enough to be implemented is hard.

Implementation wise, this would functionally have to be some language keyword (probably `readonly`) that causes a corresponding attribute to be emitted in metadata (just like C++/CLI did with `IsConst`) and having the compiler surface errors if you try to write/mutate such a pointer. An analyzer "should" be able to do this you just wouldn't get the keyword support and there are probably various edge cases and other scenarios that would have to be worked through.

**Reply by msedi** [reply]
@Joe4evr + @tannergooding : In the end thats my solution now. Not having the `Span<T>`as a field but having a `Get`method. I must admit I find this even better than having the `void*`;-)

**Reply by Joe4evr** [reply]
> Implementation wise, this would functionally have to be some language keyword (probably `readonly`) that causes a corresponding attribute to be emitted in metadata

I'm just curious where you envision this keyword to sit, because if it *is* `readonly`, most places I can think of to put it it's already legal.

**Reply by tannergooding** [reply]
The same place it goes for `ref readonly` which has the same issue ;)

But as I said, I don't expect this would actually be implemented, its a very niche scenario that ends up being just "nice to have" with decently existing alternatives for some sceanrios.
